### PR TITLE
Improve error message description

### LINKED ISSUE #70: [FEATURE] Enhance error message descriptions
The error logging in the JDBC driver and SQL-Gremlin has some gaps. Enhancing this would provide better debugging going forward - need to distinguish between supported an not supported errors.

### PR DESCRIPTION
**Pull request by simonz-bq**
### Summary

Improve error message description

### Description

Distinguished between some supported vs not supported errors.

### Related Issue

https://bitquill.atlassian.net/browse/AN-909

### Additional Reviewers
@birschick-bq
@lyndonb-bq
@xiazcy
@simonz-bq
@alexey-temnikov
@kylepbit

### CODE REVIEW

File: sql-gremlin/src/main/java/org/twilmes/sql/gremlin/adapter/util/SQLNotSupportedException.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ */
+public class SQLNotSupportedException extends SQLException {
+
+    /**
+     * SQLNotSupportedException constructor.
+     *
+     * @param message   Message of the exception.
+     * @param cause     Underlying cause for the exception.
+     */
+    public SQLNotSupportedException(String message, Throwable cause) {
```
Comment (lyndonbauto) [inline sql-gremlin/src/main/java/org/twilmes/sql/gremlin/adapter/util/SQLNotSupportedException.java]:
```suggestion
    public SQLNotSupportedException(final String message, final Throwable cause) {
```

File: sql-gremlin/src/main/java/org/twilmes/sql/gremlin/adapter/util/SQLNotSupportedException.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        super(message, cause);
+    }
+
+    /**
+     * SQLNotSupportedException constructor.
+     *
+     * @param message   Message of the exception.
+     */
+    public SQLNotSupportedException(String message) {
```
Comment (lyndonbauto) [inline sql-gremlin/src/main/java/org/twilmes/sql/gremlin/adapter/util/SQLNotSupportedException.java]:
```suggestion
    public SQLNotSupportedException(final String message) {
```

File: sql-gremlin/src/main/java/org/twilmes/sql/gremlin/adapter/util/SqlGremlinError.java [hunk 3]
```diff
@@ -117,17 +118,17 @@ private static String getMessage(final SqlGremlinError key, final Object... form
...
      */
     public static SQLException create(final SqlGremlinError key, final Logger logger, final Throwable cause,
-                                      final Object... formatArgs) {
+                                      final boolean isNotSupported, final Object... formatArgs) {
         final String message = getMessage(key, formatArgs);
         final SQLException exception;
 
         if (cause == null) {
-            exception = new SQLException(message);
+            exception = isNotSupported? new SQLNotSupportedException(message) : new SQLException(message);
```
Comment (lyndonbauto) [inline sql-gremlin/src/main/java/org/twilmes/sql/gremlin/adapter/util/SqlGremlinError.java]:
```suggestion
            exception = isNotSupported ? new SQLNotSupportedException(message) : new SQLException(message);
```

File: sql-gremlin/src/main/java/org/twilmes/sql/gremlin/adapter/util/SqlGremlinError.java [hunk 4]
```diff
@@ -117,17 +118,17 @@ private static String getMessage(final SqlGremlinError key, final Object... form
...
             if (logger != null) {
                 logger.error(message);
             }
         } else {
-            exception = new SQLException(message, cause);
+            exception = isNotSupported? new SQLNotSupportedException(message, cause) : new SQLException(message, cause);
```
Comment (lyndonbauto) [inline sql-gremlin/src/main/java/org/twilmes/sql/gremlin/adapter/util/SqlGremlinError.java]:
```suggestion
            exception = isNotSupported ? new SQLNotSupportedException(message, cause) : new SQLException(message, cause);
```

File: sql-gremlin/src/test/java/org/twilmes/sql/gremlin/adapter/GremlinSqlNotSupportedTest.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    @Test
+    public void testSubQuery() {
+        // Sub Query testing = currently caught by generic catch-all
+        runQueryTestThrows("Select name FROM Person WHERE age = (SELECT age FROM person WHERE name = 'Tom')",
+                "Error: Unknown operator: %s.");
+    }
+
+    // TODO: Handle CAST
```
Comment (lyndonbauto) [inline sql-gremlin/src/test/java/org/twilmes/sql/gremlin/adapter/GremlinSqlNotSupportedTest.java:52]: What's going on here? Does it not throw appropriately?

Comment (simonz-bq) [reply sql-gremlin/src/test/java/org/twilmes/sql/gremlin/adapter/GremlinSqlNotSupportedTest.java:52]: Yeah. It throws a NPE and I made a ticket for this.

File: sql-gremlin/src/test/java/org/twilmes/sql/gremlin/adapter/GremlinSqlNotSupportedTest.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        // OFFSET testing - currently not implemented.
+        runNotSupportedQueryTestThrows("SELECT name FROM person OFFSET 1",
+                "Unsupported: OFFSET is not currently supported.");
+
+    }
+
+    @Test
+    public void testSubQuery() {
+        // Sub Query testing = currently caught by generic catch-all
+        runQueryTestThrows("Select name FROM Person WHERE age = (SELECT age FROM person WHERE name = 'Tom')",
```
Comment (lyndonbauto) [inline sql-gremlin/src/test/java/org/twilmes/sql/gremlin/adapter/GremlinSqlNotSupportedTest.java:48]: This error message isn't great - Would be better if it said something about embedded SELECT messages are not supported. I wonder if we can detect this and throw an exception earlier.

Comment (simonz-bq) [reply sql-gremlin/src/test/java/org/twilmes/sql/gremlin/adapter/GremlinSqlNotSupportedTest.java:48]: What happens is `Caused by: java.sql.SQLException: Error: Unknown operator: SCALAR_QUERY.`

Thrown by `GremlinSqlFactory.createOperator`

File: sql-gremlin/src/test/java/org/twilmes/sql/gremlin/adapter/GremlinSqlNotSupportedTest.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    @Override
+    protected DataSet getDataSet() {
+        return DataSet.SPACE;
+    }
+
+    @Test
+    public void testOffset() throws SQLException {
+        // OFFSET testing - currently not implemented.
+        runNotSupportedQueryTestThrows("SELECT name FROM person OFFSET 1",
+                "Unsupported: OFFSET is not currently supported.");
```
Comment (lyndonbauto) [inline sql-gremlin/src/test/java/org/twilmes/sql/gremlin/adapter/GremlinSqlNotSupportedTest.java]: Can we update the underlying function to do a String comparison on the error message & exception message?

Additionally, can we use the bundled error messages in these tests?

Comment (simonz-bq) [reply sql-gremlin/src/test/java/org/twilmes/sql/gremlin/adapter/GremlinSqlNotSupportedTest.java]: Done

**Review by lyndonbauto — APPROVED** [review]
Approved with acknowledgement that this cast test issue is logged and a solution will be provided in a separate ticket.

**Pull request merged**